Add new features: Safe LoRA

chiayi-hsu opened the pull request:
Hello,

We have published a paper called Safe LoRA (https://arxiv.org/abs/2405.16833).
This work focuses on improving the safety of well-trained LoRA models.
Additionally, I have provided an example implementation in the model.py file to illustrate how to apply the Safe LoRA approach effectively.

BenjaminBossan commented:
Thanks for proposing this PR to add the Safe LoRA method to PEFT. I have only skimmed the code and paper, so this is not an in-depth review yet.

Based on what I saw, this is my high level understanding. As a user, I start out by training a LoRA adapter on my dataset using an aligned model as the base model. Next, I want to restore safety which may be reduced after training. For this, I take my trained LoRA model, the base model, and the aligned model, Then Safe LoRA will create a new adapter that "injects" the alignment back into my LoRA adapter. Please correct me if my understanding is incorrect.

Implementation-wise, I wonder if we really need a `SafeLoRA` class. Here is a proposal for a different user API, LMK what you think:

```python
peft_model = ...  # my trained LoRA model

# apply_safe_lora is a new function that does what SafeLoRA currently does
# additional arguments like select_layers_type are option args for apply_safe_lora
aligned_model = apply_safe_lora(peft_model, base_model_id, aligned_model_id)

# persist the new safe LoRA adapter
aligned_model.save_pretrained(...)
```

IMO, this would be a simpler API and it would achieve the same thing at the end.

In this example, the `aligned_model` could actually be the same `peft_model` as initially, just with a new LoRA adapter loaded which is the Safe LoRA adapter.

Another concern I saw when I read the code is that it looks like it will require a lot of memory. IIUC, we need to have the PEFT model, a copy of the PEFT model, the base model, and the aligned model in memory all at once. Even on CPU RAM, this could be difficult to achieve for many users. I wonder if it would be possible to load the weights from the base model and the aligned model one at a time, at least as an option.

I have already implemented something like this for another use case. Here is the code:

https://github.com/huggingface/peft/blob/ccc350151f95a9ff95da046bae5671da75eab52f/src/peft/utils/loftq_utils.py#L263-L411

It only works for safetensors, because pickle does not allow to load the weights lazily, but I still think it could be a nice addition.

Also, do we really need a full copy of the PEFT model? Would it be sufficient to only copy the LoRA weights? Ideally, I would like to see a solution that works even if the user has less than twice the memory required for the base model. If this doesn't work, it's also okay, but it would greatly reduce the number of users who can use the method.

chiayi-hsu commented:
Thank you for taking the time to read the code and the paper.

Yes, your understanding is correct. 

Since SafeLoRA is not a new training method for LoRA, but rather a process that enhances the safety of a **well-trained** LoRA model through certain operations, it may not be necessary to have a separate SafeLoRA class.

The most CPU memory-intensive operation here is likely when executing ```get_aligned_matrix()```, as it requires loading both the base model and the aligned model. If it were possible to implement lazy loading for both models simultaneously, allowing for the subtraction of weights in the same position to obtain the aligned matrix, it could potentially be a solution.

Regarding your last question, my current code in ```projected_weighted()``` does load the complete PEFT model (aligned model + LoRA). However, it is indeed possible to operate only on the LoRA weights without needing to load the weights of the aligned model as well. Once the user has obtained the new LoRA weights, they can then add them to the original aligned model.

BenjaminBossan commented:
Thanks for confirming and further explaining. Would you be willing to make the suggested changes? I think it would help a lot with user adoption of the new method.

chiayi-hsu commented:
Yes, I will make the changes you suggested.

I would like to ask if SafeLoRA is modified into the form of a function like ```apply_safelora``` instead of being a ```safeLoRA``` class, should it still be placed under ```tuners/```, or somewhere else?

BenjaminBossan commented:
> Yes, I will make the changes you suggested.

Great, thanks.

> I would like to ask if SafeLoRA is modified into the form of a function like `apply_safelora` instead of being a `safeLoRA` class, should it still be placed under `tuners/`, or somewhere else?

Good question, I think it should be placed elsewhere. I don't have a strong opinion, how about `utils/safelora.py`?

BenjaminBossan commented:
@chiayi-hsu do you still plan on working on this?

chiayi-hsu commented:
Yes, I’ll update the new version this week.

Thanks!

Benjamin Bossan ***@***.***>於 2024年10月28日 週一，18:39寫道：

> @chiayi-hsu <https://github.com/chiayi-hsu> do you still plan on working
> on this?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/huggingface/peft/pull/2098#issuecomment-2441217354>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AS67YGV2JRHZNJTO6DFGBFLZ5YA6FAVCNFSM6AAAAABO3KQJSKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDINBRGIYTOMZVGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

chiayi-hsu commented:
Hello, 

I have changed the ```SafeLoRA```` class into the ```apply_safelora```` function in the ````peft/utils/safelora.py````. For loading the base and aligned models, based on your suggestion, I now use ```safetensors``` to load the weights.

Additionally, regarding the projected weights, I originally loaded the weights of both the complete base model and the PEFT model, but I have now modified it to only load the weights of the PEFT model. I think this should be more memory-efficient for the user's GPU/CPU.

If you have any suggestions, please let me know. 

Thank you!

BenjaminBossan commented on src/peft/utils/safelora.py:
```diff
@@ -0,0 +1,249 @@
+class SafeLoRAConfig:
+    """
+    This is the configuration class to store the configuration of a safeLoRA.
+    """
```
Could you please also document the parameters in the docstring? You can just use the same description as below in the help.

BenjaminBossan commented on src/peft/utils/safelora.py:
```diff
@@ -0,0 +1,249 @@
+# Copyright 2023-present the HuggingFace Inc. team.
```
```suggestion
# Copyright 2024-present the HuggingFace Inc. team.
```

BenjaminBossan commented on src/peft/utils/safelora.py:
```diff
@@ -0,0 +1,249 @@
+    """
+
+    base_model_path: str = field(
+        default=None,
```
Let's not provide any defaults if `None` is not a valid argument.

BenjaminBossan commented on src/peft/utils/safelora.py:
```diff
@@ -0,0 +1,249 @@
+    )
+
+    aligned_model_path: str = field(
+        default=None,
```
Let's not provide any defaults if `None` is not a valid argument.

BenjaminBossan commented on src/peft/utils/safelora.py:
```diff
@@ -0,0 +1,249 @@
+    )
+
+    peft_model_path: str = field(
+        default=None,
```
Let's not provide any defaults if `None` is not a valid argument.

BenjaminBossan commented on src/peft/utils/safelora.py:
```diff
@@ -0,0 +1,249 @@
+            raise ValueError("peft_model_path cannot be None.")
+
+
+class SafetensorLoader:
```
AFAICT, this class is very similar to `loftq_utils._SafetensorLoader`. Instead of copying it with small modifications, would it be possible to re-use the same class? We can move it elsewhere so that it can be more easily imported.

If this proves difficult, please let me know which parts you needed to change and why, and I can help with modifying the class to work in both cases.

chiayi-hsu replied:
I want this function to behave like the ```.from_pretrained``` method, where if the user provides a repository name, it will automatically download the files if they are not available locally, and if the user provides a local path, it will load the model directly. However, in the current function, the input needs to be a PEFT model, and it retrieves the base model from the PEFT model's config. Since my input is not a PEFT model, this causes an error.

https://github.com/huggingface/peft/blob/8eeae0a63f28a80ef42b6bfc977c34ffc0b650ff/src/peft/utils/loftq_utils.py#L262C1-L282C27

BenjaminBossan replied:
I see. I think this can be fixed. Could you try if this change works:

```diff
modified   src/peft/utils/loftq_utils.py
@@ -269,8 +269,15 @@ class _SafetensorLoader:
 
     def __init__(self, peft_model, model_path):
         if model_path is None:
+            if isinstance(peft_model, str):
+                name_or_path = peft_model
+                base_model_prefix = None
+            else:
+                name_or_path = peft_model
+                base_model_prefix = getattr(peft_model.get_base_model(), "base_model_prefix", None)
+            base_model_prefix = getattr(peft_model.get_base_model(), "base_model_prefix", None)
             try:
-                model_path = snapshot_download(peft_model.base_model.config._name_or_path, local_files_only=True)
+                model_path = snapshot_download(name_or_path, local_files_only=True)
             except (AttributeError, HFValidationError) as exc:
                 raise ValueError(
                     "The provided model does not appear to be a transformers model or is a local model. In this case, "
@@ -286,7 +293,7 @@ class _SafetensorLoader:
             model_path = os.path.join(model_path, suffix)
 
         self.model_path = model_path
-        self.base_model_prefix = getattr(peft_model.get_base_model(), "base_model_prefix", None)
+        self.base_model_prefix = base_model_prefix
         self.prefix = "base_model.model."
         self.is_sharded = False
         self.weight_map = None
```

This way, if the first argument is just a HF model id, it should also work. If this works, we should also rename the arguments, but that can be decided later.

chiayi-hsu replied:
I tried the modified version and changed the last line shown below, but it still has issues.

```diff

modified   src/peft/utils/loftq_utils.py
@@ -269,8 +269,15 @@ class _SafetensorLoader:
 
     def __init__(self, peft_model, model_path):
         if model_path is None:
+           if isinstance(peft_model, str):
+                name_or_path = peft_model
+                base_model_prefix = None
+            else:
+                name_or_path = peft_model
+                base_model_prefix = getattr(peft_model.get_base_model(), "base_model_prefix", None)
-            base_model_prefix = getattr(peft_model.get_base_model(), "base_model_prefix", None)
```

1. When the first argument is repo_id (HF model id), the error message is ```LocalEntryNotFoundError: Cannot find an appropriate cached snapshot folder for the specified revision on the local disk and outgoing traffic has been disabled. To enable repo look-ups and downloads online, pass 'local_files_only=False' as input```. To enable automatic model downloading, change ```local_files_only=True``` to ```False``` in ```snapshot_download(name_or_path, local_files_only=True)```.
2. When the first argument is a local path instead of a repo id (HF model id), it cannot pass the section shown below and goes into an exception.  I think it may be necessary to check if the path exists on disk, and only if it doesn’t, proceed with ```snapshot_download```.
```python
    try:
        model_path = snapshot_download(name_or_path,local_files_only=True)
    except (AttributeError, HFValidationError) as exc:
        raise ValueError(
            "The provided model does not appear to be a transformers model or is a local model. In this case, "
            "you must pass the model_path argument that points to the safetensors file."
        ) ```

BenjaminBossan replied:
I see, thanks for testing. How about this: We make `local_files_only` an argument to the class. By default, it is `True`, so same as now, but the caller can pass `False`, which would allow the model to be downloaded. Then `apply_safelora` can also take a `local_files_only` argument, which defaults to `False`, and it is passed down to `_SafetensorLoader`. WDYT?

chiayi-hsu replied:
I think that sounds feasible and there's no issue with it.

Regarding the two points I mentioned above, the user is inputting a local path instead of a repo_id, so the necessary modifications may be as follows:
```diff

class _SafetensorLoader:
+    def __init__(self, peft_model, model_path=None, local_files_only=True):
        if model_path is None:
            if isinstance(peft_model, str):
                name_or_path = peft_model
                base_model_prefix = None
            else:
                name_or_path = peft_model
                base_model_prefix = getattr(peft_model.get_base_model(), "base_model_prefix", None)
+           if os.path.exists(name_or_path):
+                model_path = name_or_path
+           else:
                try:
+                    model_path = snapshot_download(name_or_path,local_files_only=local_files_only)
                except (AttributeError, HFValidationError) as exc:
                    raise ValueError(
                        "The provided model does not appear to be a transformers model or is a local model. In this case, "
                        "you must pass the model_path argument that points to the safetensors file."
                    ) 

        suffix = "model.safetensors"
        if not model_path.endswith(suffix):
            model_path = os.path.join(model_path, suffix)

```

This version works for both the local path and repo_id. 
If you have any suggestions, please let me know.

BenjaminBossan replied:
This sounds good. Let's just rename the `peft_model` argument to `peft_model_or_model_id`. It's a bit ugly but otherwise the name would be confusing.

BenjaminBossan commented on src/peft/utils/safelora.py:
```diff
@@ -0,0 +1,249 @@
+    ]
+    v = []
+    for name_base, name_align in zip(base, align):
+        assert (
```
Let's raise a proper `ValueError`.

BenjaminBossan commented on src/peft/utils/safelora.py:
```diff
@@ -0,0 +1,249 @@
+    align = [
+        name for name in sl_align.weight_map.keys() if any(v in name for v in list(peft_config["target_modules"]))
+    ]
+    v = []
```
Let's use a more descriptive variable name.

BenjaminBossan commented on src/peft/utils/safelora.py:
```diff
@@ -0,0 +1,249 @@
+    i = 0
+    dis = []
+    cos_total = []
+    for name_A, name_B in zip(names_A, names_B):
```
Can we use `enumerate` and then we don't need the separate `idx` variable?

BenjaminBossan commented on src/peft/utils/safelora.py:
```diff
@@ -0,0 +1,249 @@
+    names_A = [name for name in peft_weights.keys() if "lora_A" in name]
+    names_B = [name for name in peft_weights.keys() if "lora_B" in name]
+    idx = 0
+    i = 0
```
Let's give this variable a proper name.

BenjaminBossan commented on src/peft/utils/safelora.py:
```diff
@@ -0,0 +1,249 @@
+        fW = torch.mm(W, A)
+        ori = torch.mm(ori_peft_weights[name_B], A)
+
+        cos = numpy.round(torch.nn.functional.cosine_similarity(fW.reshape(1, -1), ori.reshape(1, -1)).item(), 5)
```
Can we use torch instead of numpy?

BenjaminBossan commented on src/peft/utils/safelora.py:
```diff
@@ -0,0 +1,249 @@
+
+def apply_safelora(configs):
+    """
+    ===================================================
```
Let's remove the `=`. Also, the first line should be a description of what this function does. You can also include a link to the paper.

Regarding the syntax for examples, check for instance here: https://github.com/huggingface/peft/blob/8eeae0a63f28a80ef42b6bfc977c34ffc0b650ff/src/peft/peft_model.py#L1404-L1429

BenjaminBossan commented on src/peft/utils/safelora.py:
```diff
@@ -0,0 +1,249 @@
+    return peft_weights, cos_total
+
+
+def apply_safelora(configs):
```
Let's add a type annotation for `SafeLoRAConfig`. Also, let's rename the variable `safelora_config`.

BenjaminBossan reviewed (changes requested):
Thanks for the updates. I reviewed the code and left a couple of comments. Before pushing your changes, always make sure to run `make style`.

When this is addressed, there will be a few more steps before we can merge the PR:

1. We should add an example to the `examples/` folder. Ideally, this can be something similar to what's in the paper, so that we can replicate the results there.
2. This feature should be documented.

Without these steps, users will have a hard time finding this feature.

chiayi-hsu commented:
Sorry for the inconvenience. The pull request was closed due to syncing with the latest version of PEFT. I have resubmitted the PR.
